src/ray/gcs/gcs_client/gcs_client.cc

```diff
@@ -145,36 +146,61 @@ std::pair<std::string, int> GcsClient::GetGcsServerAddress() const {
 
 PythonGcsClient::PythonGcsClient(const GcsClientOptions &options) : options_(options) {}
 
-Status PythonGcsClient::Connect() {
+namespace {
+Status HandleGcsError(rpc::GcsStatus status) {
+  RAY_CHECK(status.code() != static_cast<int>(StatusCode::OK));
+  return Status::Invalid(status.message() +
+                         " [GCS status code: " + std::to_string(status.code()) + "]");
+}
+}  // namespace
+
+Status PythonGcsClient::Connect(const ClusterID &cluster_id, int64_t timeout_ms) {
   channel_ =
       rpc::GcsRpcClient::CreateGcsChannel(options_.gcs_address_, options_.gcs_port_);
   kv_stub_ = rpc::InternalKVGcsService::NewStub(channel_);
   runtime_env_stub_ = rpc::RuntimeEnvGcsService::NewStub(channel_);
   node_info_stub_ = rpc::NodeInfoGcsService::NewStub(channel_);
   job_info_stub_ = rpc::JobInfoGcsService::NewStub(channel_);
   autoscaler_stub_ = rpc::autoscaler::AutoscalerStateService::NewStub(channel_);
-  return Status::OK();
-}
 
-Status HandleGcsError(rpc::GcsStatus status) {
-  RAY_CHECK(status.code() != static_cast<int>(StatusCode::OK));
-  return Status::Invalid(status.message() +
-                         " [GCS status code: " + std::to_string(status.code()) + "]");
-}
-
-void GrpcClientContextWithTimeoutMs(grpc::ClientContext &context, int64_t timeout_ms) {
-  if (timeout_ms != -1) {
-    context.set_deadline(std::chrono::system_clock::now() +
-                         std::chrono::milliseconds(timeout_ms));
+  if (cluster_id.IsNil()) {
+    RAY_LOG(INFO) << "Retrieving cluster ID from GCS server.";
+    grpc::ClientContext context;
+    PrepareContext(context, timeout_ms);
+
+    rpc::GetClusterIdRequest request;
+    rpc::GetClusterIdReply reply;
+
+    auto status =
+        GrpcStatusToRayStatus(node_info_stub_->GetClusterId(&context, request, &reply));
+    while (!status.IsTimedOut()) {
```

Comment:
Should we fail after several retries?

Reply:
Looks like all the other methods use timeouts

Reply:
Added retries to emulate the @_auto_reconnect on the python side